src/Experiment/src/AssemblyBuilder.cs

```diff
@@ -7,21 +7,69 @@
+using System.Runtime.CompilerServices;
 
 namespace System.Reflection.Emit.Experimental
 {
+
     public class AssemblyBuilder: System.Reflection.Assembly
     {
-        internal AssemblyBuilder() { }
+        private static readonly Guid _s_guid = new Guid("87D4DBE1-1143-4FAD-AAB3-1001F92068E6");//Some random ID, Need to look into how these should be generated.
+        private static readonly BlobContentId _s_contentId = new BlobContentId(_s_guid, 0x04030201);
+        private BlobBuilder _emptyBlob = new BlobBuilder();
+        public override string? FullName { get; }
+        private static MetadataBuilder _metadata = new MetadataBuilder();
+        private static IDictionary<string,ModuleBuilder> moduleStorage = new Dictionary<string, ModuleBuilder>();
+        public AssemblyBuilder() { }
+        private AssemblyBuilder(string name) 
+        {
+            FullName = name;
+        }
 
         public void Save(string assemblyFileName)
         {
-            //New method
+            if (assemblyFileName == null)
+            {
+                throw new ArgumentNullException();
+            }
+            if (FullName == null)
+            {
+                throw new ArgumentNullException();
+            }
+            //Add assembly metadata
+            _metadata.AddAssembly(//Metadata is added for the new assembly - Current design - metdata generated only when Save method is called.
+               _metadata.GetOrAddString(value: FullName),//FullName, CultureName?
+               version: new Version(1, 0, 0, 0),
+               culture: default(StringHandle),
+               publicKey: default(BlobHandle),
+               flags: 0,
+               hashAlgorithm: AssemblyHashAlgorithm.None);
+            //Add each module's medata
+            foreach (KeyValuePair<string, ModuleBuilder> entry in moduleStorage)
+            {
+                entry.Value.AppendMetadata(_metadata);
+            }
+            using var peStream = new FileStream(assemblyFileName, FileMode.OpenOrCreate, FileAccess.ReadWrite);
+            var ilBuilder = new BlobBuilder();
+            WritePEImage(peStream, _metadata, ilBuilder);
+            peStream.Dispose();//close or dispose?
         }
 
-        public static System.Reflection.Emit.AssemblyBuilder DefineDynamicAssembly(System.Reflection.AssemblyName name, System.Reflection.Emit.AssemblyBuilderAccess access) 
-        { 
-            throw new NotImplementedException(); 
+        public static System.Reflection.Emit.Experimental.AssemblyBuilder DefineDynamicAssembly(System.Reflection.AssemblyName name, System.Reflection.Emit.AssemblyBuilderAccess access)
+        {
+            if (name == null || name.Name == null)
+            {
+                throw new ArgumentNullException();
+            }
+            //AssemblyBuilderAccess affects runtime managment only and is not relevant for saving to disk.
+            AssemblyBuilder currentAssembly = new AssemblyBuilder(name.Name);
```

Comment:
I would thread the `AssemblyName` into the `AssemblyBuilder`. As @jkotas has mentioned you need those details during the save routine.